File: app/src/main/java/org/secuso/privacyfriendlyminesweeper/activities/adapter/PlayRecyclerViewAdapter.java
package org.secuso.privacyfriendlyminesweeper.activities.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.TextView;
import android.os.Bundle;
import org.secuso.privacyfriendlyminesweeper.R;
import org.secuso.privacyfriendlyminesweeper.activities.helper.BaseActivity;
import org.secuso.privacyfriendlyminesweeper.activities.helper.CellView;

public class PlayRecyclerViewAdapter extends RecyclerView.Adapter<PlayRecyclerViewAdapter.ViewHolder> {

    private int[] mData;
    private LayoutInflater mInflater;
    private ItemClickListener mClickListener;
    private int maxHeightOfCells;
    private int maxWidthOfCells;

    // data is passed into the constructor
    public PlayRecyclerViewAdapter(Context context, int[] data) {
        this.mInflater = LayoutInflater.from(context);
        this.mData = data;
    }

    // data is passed into the constructor
    public PlayRecyclerViewAdapter(Context context, int[] data, int maxHeight) {
        this.mInflater = LayoutInflater.from(context);
        this.mData = data;
        maxHeightOfCells = maxHeight;
    }

    // inflates the cell layout from xml when needed
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder (@NonNull ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.playingfield_cell, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the textview in each cell
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
  //      holder.myTextView.setText(String.valueOf(mData[position]));
    }

    // total number of cells
    @Override
    public int getItemCount() {
        return mData.length;
    }

    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        CellView myTextView;

        ViewHolder(View itemView) {
            super(itemView);
            myTextView = (CellView) itemView.findViewById(R.id.cell);
            itemView.setOnClickListener(this);
            myTextView.setTag(maxHeightOfCells);
   //         myTextView.setTag(R.id.width ,maxWidthOfCells);
        }

        @Override
        public void onClick(View view) {
            if (mClickListener != null) mClickListener.onItemClick(view, getAdapterPosition());
        }
    }

    // convenience method for getting data at click position
    public int getItem(int id) {
        return mData[id];
    }

    // allows clicks events to be caught
    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, int position);
    }
}

File: app/src/main/java/org/secuso/privacyfriendlyminesweeper/activities/adapter/PlayRecyclerViewAdapter.kt
package org.secuso.privacyfriendlyminesweeper.activities.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.secuso.privacyfriendlyminesweeper.R
import org.secuso.privacyfriendlyminesweeper.activities.helper.CellView
import org.secuso.privacyfriendlyminesweeper.activities.model.grid.MinesweeperCell

/**
 * @author max-dreger, Patrick Schneider
 * @version 20221218
 * This class implements an adapter to handle the views of the cells on the playing field
 */
class PlayRecyclerViewAdapter(
    val context: Context,
    val layoutInflater: LayoutInflater,
    val data: Array<MinesweeperCell>,
    val maxHeight: Int,
    val size: Int,
    val colors: IntArray,
    val backgroundColor: Int,
    val onItemClickListener: ItemClickListener
) : RecyclerView.Adapter<PlayRecyclerViewAdapter.ViewHolder?>() {

    // inflates the cell layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = layoutInflater.inflate(R.layout.playingfield_cell, parent, false)
        view.findViewById<View?>(R.id.playingfield_row).setLayoutParams(LinearLayout.LayoutParams(size, size))
        view.findViewById<View?>(R.id.playingfield_row).requestLayout()

        return PlayRecyclerViewAdapter.ViewHolder(view, maxHeight, onItemClickListener)
    }

    // binds the data to the textview in each cell
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val cellBackground = holder.itemView.findViewById<TextView>(R.id.cell_background)
        val cellForeground = holder.itemView.findViewById<TextView>(R.id.cell_decoration)
        val cell = data[position]

        when (cell.state) {
            MinesweeperCell.State.FLAGGED -> cellForeground.setBackgroundResource(R.drawable.flagge)
            MinesweeperCell.State.REVEALED -> if (cell.isBomb) {
                cellForeground.setBackgroundResource(R.drawable.mine_x)
            } else {
                if (cell.bombs == 0) {
                    cellForeground.setText("")
                } else {
                    cellForeground.setText(cell.bombs.toString())
                }
                cellForeground.setTextColor(colors[cell.bombs])
                cellForeground.setBackgroundResource(android.R.color.transparent)
                cellBackground.setBackgroundColor(backgroundColor)
            }

            MinesweeperCell.State.COVERED -> cellForeground.setBackgroundResource(android.R.color.transparent)
        }
    }

    // total number of cells
    override fun getItemCount(): Int {
        return data.size
    }

    // stores and recycles views as they are scrolled off screen
    class ViewHolder internal constructor(itemView: View, val maxHeight: Int, val onItemClickListener: ItemClickListener? = null) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var myTextView: CellView

        init {
            myTextView = itemView.findViewById<View?>(R.id.cell) as CellView
            itemView.setOnClickListener(this)
            myTextView.setTag(maxHeight)
        }

        override fun onClick(view: View?) {
            onItemClickListener?.onItemClick(view, getAdapterPosition())
        }
    }

    // convenience method for getting data at click position
    fun getItem(index: Int): MinesweeperCell? {
        return data[index]
    }

    fun setItem(index: Int, cell: MinesweeperCell?) {
        data[index] = cell!!
    }

    // parent activity will implement this method to respond to click events
    interface ItemClickListener {
        fun onItemClick(view: View?, position: Int)
    }
}
